<a href="https://colab.research.google.com/github/Mashi25/Red-Wine-Quality-Analysis/blob/main/Red_wine_quality.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>



Red Wine Quality

G Nafisa Mashitha

#REGRESSION

##Preprocessing the data

Importing relevant libraries required for the project.

In [ ]:
import pandas as pd
import numpy as np

Reading the dataset.

In [ ]:
dataset= pd.read_csv('/content/winequality-red.csv')

In [ ]:
dataset

fixed acidity  volatile acidity  citric acid  ...  sulphates  alcohol  quality
0               7.4             0.700         0.00  ...       0.56      9.4        5
1               7.8             0.880         0.00  ...       0.68      9.8        5
2               7.8             0.760         0.04  ...       0.65      9.8        5
3              11.2             0.280         0.56  ...       0.58      9.8        6
4               7.4             0.700         0.00  ...       0.56      9.4        5
...             ...               ...          ...  ...        ...      ...      ...
1594            6.2             0.600         0.08  ...       0.58     10.5        5
1595            5.9             0.550         0.10  ...       0.76     11.2        6
1596            6.3             0.510         0.13  ...       0.75     11.0        6
1597            5.9             0.645         0.12  ...       0.71     10.2        5
1598            6.0             0.310         0.47  ...       0.66     11.0        6

[1599 rows x 12 columns]

To check the datatype of all the features:

In [ ]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


To understand the relation between various features and the quality we find the Correlation. 

In [ ]:
dataset.corr()['quality']

fixed acidity           0.124052
volatile acidity       -0.390558
citric acid             0.226373
residual sugar          0.013732
chlorides              -0.128907
free sulfur dioxide    -0.050656
total sulfur dioxide   -0.185100
density                -0.174919
pH                     -0.057731
sulphates               0.251397
alcohol                 0.476166
quality                 1.000000
Name: quality, dtype: float64

Splitting the dataset into X and y:

Here x are all the features and y is target. 

In [ ]:
X=dataset.iloc[:,:-1].values
y=dataset.iloc[:,-1].values

In [ ]:
X

array([[ 7.4  ,  0.7  ,  0.   , ...,  3.51 ,  0.56 ,  9.4  ],
       [ 7.8  ,  0.88 ,  0.   , ...,  3.2  ,  0.68 ,  9.8  ],
       [ 7.8  ,  0.76 ,  0.04 , ...,  3.26 ,  0.65 ,  9.8  ],
       ...,
       [ 6.3  ,  0.51 ,  0.13 , ...,  3.42 ,  0.75 , 11.   ],
       [ 5.9  ,  0.645,  0.12 , ...,  3.57 ,  0.71 , 10.2  ],
       [ 6.   ,  0.31 ,  0.47 , ...,  3.39 ,  0.66 , 11.   ]])

In [ ]:
y

array([5, 5, 5, ..., 6, 5, 6])

Splitting the data into Test and Train Samples:

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X_train , X_test ,y_train , y_test= train_test_split(X,y, test_size=0.2)

##Multiple Linear Regression

Importing the library required to perform Multiple Linear Regression

In [ ]:
from sklearn.linear_model import LinearRegression

In [ ]:
mreg= LinearRegression()

Training the train samples:

In [ ]:
mreg.fit(X_train,y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

To predict the quality values for the Test Features:

In [ ]:
pred=mreg.predict(X_test)

To view all the predicted outputs and  the difference between the predicted and real output:

In [ ]:
for i in range(len(X_test)):
  print(X_test[i],y_test[i],pred[i],y_test[i]-pred[i])

[ 7.5      0.31     0.41     2.4      0.065   34.      60.       0.99492
  3.34     0.85    11.4    ] 6 6.33623047423467 -0.3362304742346698
[ 8.4      0.62     0.12     1.8      0.072   38.      46.       0.99504
  3.38     0.89    11.8    ] 6 6.2610557491968635 -0.2610557491968635
[ 5.6      0.605    0.05     2.4      0.073   19.      25.       0.99258
  3.56     0.55    12.9    ] 5 6.2180959099299535 -1.2180959099299535
[10.3    0.59   0.42   2.8    0.09  35.    73.     0.999  3.28   0.7
  9.5  ] 6 5.31577150344664 0.6842284965533603
[ 7.     0.45   0.34   2.7    0.082 16.    72.     0.998  3.55   0.6
  9.5  ] 5 5.195489868003314 -0.19548986800331392
[8.200e+00 6.000e-01 1.700e-01 2.300e+00 7.200e-02 1.100e+01 7.300e+01
 9.963e-01 3.200e+00 4.500e-01 9.300e+00] 5 5.047038980918687 -0.0470389809186873
[8.100e+00 7.850e-01 5.200e-01 2.000e+00 1.220e-01 3.700e+01 1.530e+02
 9.969e-01 3.210e+00 6.900e-01 9.300e+00] 5 4.722320876805917 0.2776791231940834
[9.000e+00 4.400e-01 4.900e-01 2.

R-squared is a goodness-of-fit measure for linear regression models. This is indicated in percentage.

Root Mean Square Error (RMSE) is the standard deviation of prediction errors.

In [ ]:
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error

In [ ]:
mulrmse= np.sqrt(mean_squared_error(y_test,pred))
mulscore= r2_score(y_test,pred)

In [ ]:
print('r2 score ='+ str(mulscore)+' rmse='+str(mulrmse))

r2 score =0.35477507219355753 rmse=0.6522422729888263


##Decision Tree

Importing the required library:

In [ ]:
from sklearn.tree import DecisionTreeRegressor

In [ ]:
dreg= DecisionTreeRegressor()

In [ ]:
dreg.fit(X_train,y_train)

DecisionTreeRegressor(ccp_alpha=0.0, criterion='mse', max_depth=None,
                      max_features=None, max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, presort='deprecated',
                      random_state=None, splitter='best')

In [ ]:
pred= dreg.predict(X_test)

In [ ]:
for i in range(len(X_test)):
  print(y_test[i],pred[i], pred[i]-y_test[i])

6 5.0 -1.0
6 6.0 0.0
5 5.0 0.0
6 6.0 0.0
5 5.0 0.0
5 5.0 0.0
5 5.0 0.0
5 5.0 0.0
4 7.0 3.0
7 6.0 -1.0
5 5.0 0.0
6 6.0 0.0
7 7.0 0.0
4 5.0 1.0
6 7.0 1.0
5 5.0 0.0
7 8.0 1.0
6 6.0 0.0
5 4.0 -1.0
5 6.0 1.0
5 5.0 0.0
6 6.0 0.0
7 8.0 1.0
5 5.0 0.0
7 7.0 0.0
4 5.0 1.0
5 7.0 2.0
5 5.0 0.0
6 5.0 -1.0
5 5.0 0.0
6 6.0 0.0
5 5.0 0.0
6 6.0 0.0
5 5.0 0.0
4 5.0 1.0
5 5.0 0.0
5 5.0 0.0
6 6.0 0.0
6 6.0 0.0
6 6.0 0.0
6 6.0 0.0
5 5.0 0.0
5 5.0 0.0
7 7.0 0.0
5 5.0 0.0
5 6.0 1.0
7 7.0 0.0
5 6.0 1.0
5 5.0 0.0
5 6.0 1.0
6 6.0 0.0
5 5.0 0.0
6 6.0 0.0
5 5.0 0.0
5 5.0 0.0
6 4.0 -2.0
5 5.0 0.0
6 6.0 0.0
5 5.0 0.0
6 6.0 0.0
6 6.0 0.0
6 6.0 0.0
6 6.0 0.0
7 7.0 0.0
5 5.0 0.0
6 6.0 0.0
6 7.0 1.0
5 5.0 0.0
4 5.0 1.0
5 5.0 0.0
6 5.0 -1.0
5 5.0 0.0
5 5.0 0.0
5 5.0 0.0
5 5.0 0.0
6 7.0 1.0
5 5.0 0.0
5 5.0 0.0
5 5.0 0.0
5 6.0 1.0
5 5.0 0.0
5 6.0 1.0
6 5.0 -1.0
6 6.0 0.0
6 6.0 0.0
7 7.0 0.0
5 6.0 1.0
4 6.0 2.0
6 6.0 0.0
6 6.0 0.0
6 6.0 0.0
7 6.0 -1.0
5 5.0 0.0
5 6.0 1.0
6 4.0 -2.0
5 5.0 0.0
6 5.0 -1.0
6 6.0 0.0
5 5.0 0.0


In [ ]:
dtscore= r2_score(y_test,pred)
dtrmse= np.sqrt(mean_squared_error(y_test,pred))

In [ ]:
print('r2 score ='+ str(dtscore)+' rmse='+str(dtrmse))

r2 score =0.07103501392262579 rmse=0.7826237921249264


##Random Forest Regression

Importing the library required for Random Forest Regression:

In [ ]:
from sklearn.ensemble import RandomForestRegressor

In [ ]:
rreg= RandomForestRegressor(n_estimators=100)

In [ ]:
rreg.fit(X_train,y_train)

RandomForestRegressor(bootstrap=True, ccp_alpha=0.0, criterion='mse',
                      max_depth=None, max_features='auto', max_leaf_nodes=None,
                      max_samples=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      n_estimators=100, n_jobs=None, oob_score=False,
                      random_state=None, verbose=0, warm_start=False)

In [ ]:
rfscore= r2_score(y_test,pred)
rfrmse= np.sqrt(mean_squared_error(y_test,pred))

In [ ]:
print('r2 score ='+ str(rfscore)+ ' rmse='+str(rfrmse))

r2 score =0.07103501392262579 rmse=0.7826237921249264


##Support Vector Regression

As standardising the data is required for training the data using Support Vector Regression Model, we import the required library.

Standardization is the process of putting different variables on the same scale.


In [ ]:
from sklearn.preprocessing import StandardScaler

Standard Scaler requires a two dimensional array.

As y is a one dimensional array we reshape it to a 2D format.

In [ ]:
t_train=y_train.reshape(len(y_train),1)
t_test=y_test.reshape(len(y_test),1)

In [ ]:
X_sc=StandardScaler()
y_sc=StandardScaler()

In [ ]:
X_std_train=X_sc.fit_transform(X_train)
y_std_train=y_sc.fit_transform(t_train)

In [ ]:
X_std_test=X_sc.transform(X_test)
y_std_test=y_sc.transform(t_test)

In [ ]:
from sklearn.svm import SVR

In [ ]:
sreg = SVR()

In [ ]:
sreg.fit(X_std_train,y_std_train)

/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVR(C=1.0, cache_size=200, coef0=0.0, degree=3, epsilon=0.1, gamma='scale',
    kernel='rbf', max_iter=-1, shrinking=True, tol=0.001, verbose=False)

In [ ]:
s_pred = sreg.predict(X_std_test)

In [ ]:
sv_score= r2_score(y_std_test, s_pred)
sv_rmse = np.sqrt(mean_squared_error(y_sc.inverse_transform(y_std_test), y_sc.inverse_transform(s_pred)))

In [ ]:
print('r2 score support vector ='+ str(sv_score)+' rmse support vector='+str(sv_rmse))

r2 score support vector =0.3857276589907429 rmse support vector=0.6364053973998478


##R2 score and RMSE Value

In [ ]:
print('r2 score multiplelinear ='+ str(mulscore)+' rmse multiplelinear ='+str(mulrmse))
print('r2 score decision tree  ='+ str(dtscore)+' rmse decision tree   ='+str(dtrmse))
print('r2 score random forest  ='+ str(rfscore)+ ' rmse random forest  ='+str(rfrmse))
print('r2 score support vector ='+ str(sv_score)+' rmse support vector ='+str(sv_rmse))

r2 score multiplelinear =0.35477507219355753 rmse multiplelinear =0.6522422729888263
r2 score decision tree  =0.07103501392262579 rmse decision tree   =0.7826237921249264
r2 score random forest  =0.07103501392262579 rmse random forest  =0.7826237921249264
r2 score support vector =0.3857276589907429 rmse support vector =0.6364053973998478


As Support Vector Regression shows highest percentage of R-Squared value and least RMSE error it is the best fit Model to predict the Quality of the Red Wine in this dataset. 

#CLASSIFIER

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
dataset= pd.read_csv('/content/winequality-red.csv') 

In [ ]:
dataset

fixed acidity  volatile acidity  citric acid  ...  sulphates  alcohol  quality
0               7.4             0.700         0.00  ...       0.56      9.4        5
1               7.8             0.880         0.00  ...       0.68      9.8        5
2               7.8             0.760         0.04  ...       0.65      9.8        5
3              11.2             0.280         0.56  ...       0.58      9.8        6
4               7.4             0.700         0.00  ...       0.56      9.4        5
...             ...               ...          ...  ...        ...      ...      ...
1594            6.2             0.600         0.08  ...       0.58     10.5        5
1595            5.9             0.550         0.10  ...       0.76     11.2        6
1596            6.3             0.510         0.13  ...       0.75     11.0        6
1597            5.9             0.645         0.12  ...       0.71     10.2        5
1598            6.0             0.310         0.47  ...       0.66     11.0        6

[1599 rows x 12 columns]

In [ ]:
X = dataset.iloc[:,:-1].values
y= dataset.iloc[:,-1].values

In [ ]:
X

array([[ 7.4  ,  0.7  ,  0.   , ...,  3.51 ,  0.56 ,  9.4  ],
       [ 7.8  ,  0.88 ,  0.   , ...,  3.2  ,  0.68 ,  9.8  ],
       [ 7.8  ,  0.76 ,  0.04 , ...,  3.26 ,  0.65 ,  9.8  ],
       ...,
       [ 6.3  ,  0.51 ,  0.13 , ...,  3.42 ,  0.75 , 11.   ],
       [ 5.9  ,  0.645,  0.12 , ...,  3.57 ,  0.71 , 10.2  ],
       [ 6.   ,  0.31 ,  0.47 , ...,  3.39 ,  0.66 , 11.   ]])

In [ ]:
y

array([5, 5, 5, ..., 6, 5, 6])

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X_train,X_test, y_train , y_test = train_test_split(X, y, test_size=0.2)

In [ ]:
len(X_train)

1279

Importing the Standard scaler and creating the object of it. 

In [ ]:
from sklearn.preprocessing import StandardScaler

In [ ]:
sc = StandardScaler()

In [ ]:
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [ ]:
X_test

array([[ 1.21116898, -1.47378959,  1.08440497, ...,  0.12616033,
         1.2102129 ,  0.42678564],
       [-0.529963  ,  0.4101622 , -0.05926021, ..., -0.13119065,
        -0.5730386 , -0.59594425],
       [-0.41388754, -0.36558265, -0.83903193, ..., -0.64589261,
        -0.51551435, -0.40999336],
       ...,
       [ 0.86294258,  0.74262429,  0.61654194, ...,  0.38351132,
         0.28982503,  0.98463832],
       [ 1.50135764,  0.02228978,  1.13638975, ..., -1.54662105,
        -0.5730386 ,  1.17058921],
       [-1.11034033,  0.63180359, -0.31918412, ...,  1.79894172,
         0.00220382, -0.59594425]])

##Logic Regression 

importing the required library and creating an object lcla of it.

In [ ]:
from sklearn.linear_model import LogisticRegression

In [ ]:
lcla= LogisticRegression()

the dataset is trained.

In [ ]:
lcla.fit(X_train, y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [ ]:
pred=lcla.predict(X_test)

In [ ]:
for i in range(len(pred)):
  print(pred[i] , y_test[i])

7 6
5 6
5 5
6 6
5 6
5 5
5 5
5 3
5 4
5 6
6 5
5 5
6 6
6 6
6 6
5 5
6 5
6 6
7 8
5 4
5 5
7 7
5 5
5 5
5 6
5 5
6 6
6 5
6 6
5 6
5 5
5 6
5 6
5 5
6 5
5 5
6 5
5 5
5 5
5 6
5 5
7 7
6 6
6 5
6 5
5 5
5 5
6 6
6 6
5 5
5 5
7 7
5 5
5 6
5 5
6 7
5 5
5 6
5 6
6 6
5 5
5 6
5 5
5 5
6 5
6 6
5 6
6 5
5 5
6 6
6 7
5 5
6 6
6 6
6 6
5 4
7 6
6 4
7 6
6 5
5 5
5 5
5 5
6 6
6 6
6 6
6 6
6 6
7 7
6 5
6 7
6 5
7 7
7 6
5 6
5 5
7 6
5 5
5 6
6 8
5 6
6 6
5 6
6 5
7 7
5 6
5 5
5 6
5 6
5 5
5 6
6 6
6 6
5 5
6 7
7 7
7 6
5 5
6 5
5 5
6 6
5 5
5 6
7 6
5 5
5 5
5 5
6 6
6 7
6 6
7 8
5 6
7 6
5 6
5 5
5 5
6 6
5 5
6 6
5 5
6 7
6 6
6 6
6 6
5 6
5 5
6 5
6 6
6 7
5 5
6 5
6 6
7 6
6 5
5 6
6 6
6 5
6 5
6 6
6 6
5 5
6 5
5 6
7 7
6 6
5 5
6 6
5 5
6 7
6 6
6 6
6 6
6 5
5 6
6 5
5 6
6 5
6 5
6 6
5 6
5 6
5 5
5 5
6 6
5 5
5 5
6 6
5 5
6 6
5 6
5 4
6 5
5 5
6 6
6 6
6 6
5 5
6 4
5 5
5 5
6 5
6 6
5 4
5 6
5 6
7 6
5 5
5 5
5 5
5 6
6 6
5 4
5 5
5 5
5 5
6 5
5 6
5 5
5 6
5 5
6 6
6 6
5 5
5 5
5 7
6 6
6 5
5 6
5 5
6 6
5 5
6 6
6 7
6 6
5 5
7 6
5 5
6 7
5 5
7 6
6 5
5 5
5 5
5 5
6 6
5 6
5 5
7 7
5 5
5 5


Accuracy Score is the ratio of number of correct predictions to the total number of input samples. 

The Confusion matrix compares the actual target values with those predicted by the machine learning model.

importing accuracy score and confusion matrix and creating objects of it.

In [ ]:
from sklearn.metrics import accuracy_score , confusion_matrix

In [ ]:
ac_l = accuracy_score(y_test, pred)
cm_l = confusion_matrix(y_test, pred)

In [ ]:
ac_l

0.578125

In [ ]:
cm_l

array([[  0,   0,   2,   0,   0,   0],
       [  0,   0,   7,   2,   1,   0],
       [  0,   0, 104,  37,   0,   0],
       [  0,   0,  52,  71,  14,   0],
       [  0,   0,   2,  15,  10,   0],
       [  0,   0,   0,   1,   2,   0]])

##Decision Tree 

In [ ]:
from sklearn.tree import DecisionTreeClassifier

In [ ]:
dcla =  DecisionTreeClassifier()

In [ ]:
dcla.fit(X_train,y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

 Predicting the test set results

In [ ]:
pred = dcla.predict(X_test)

Making the confusion matrix

In [ ]:
from sklearn.metrics import accuracy_score, confusion_matrix

In [ ]:
cm_d = confusion_matrix(y_test, pred)
ac_d = accuracy_score(y_test, pred)

In [ ]:
cm_d

array([[ 1,  0,  1,  0,  0,  0],
       [ 0,  0,  4,  4,  2,  0],
       [ 2,  2, 95, 36,  6,  0],
       [ 0,  1, 37, 76, 22,  1],
       [ 0,  0,  2,  7, 17,  1],
       [ 0,  0,  0,  1,  2,  0]])

In [ ]:
ac_d

0.590625

##Random Forest Classification

In [ ]:
from sklearn.ensemble import RandomForestClassifier

In [ ]:
rcla = RandomForestClassifier()

In [ ]:
rcla.fit(X_train,y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [ ]:
pred = rcla.predict(X_test)

In [ ]:
from sklearn.metrics import accuracy_score , confusion_matrix

In [ ]:
ac_r =  accuracy_score(y_test , pred)
cm_r = confusion_matrix(y_test , pred)

In [ ]:
ac_r

0.68125

In [ ]:
cm_r

array([[  0,   0,   1,   1,   0,   0],
       [  0,   0,   5,   5,   0,   0],
       [  0,   0, 116,  25,   0,   0],
       [  0,   0,  35,  86,  16,   0],
       [  0,   0,   1,  10,  16,   0],
       [  0,   0,   0,   1,   2,   0]])

## K Nearest Neighbors Algorithm

Preprocess the data


In [ ]:
from sklearn.neighbors import KNeighborsClassifier

In [ ]:
kcla = KNeighborsClassifier()
kcla.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                     weights='uniform')

In [ ]:
pred = kcla.predict(X_test)

In [ ]:
from sklearn.metrics import accuracy_score, confusion_matrix

In [ ]:
cm_k = confusion_matrix(y_test, pred)
ac_k = accuracy_score(y_test, pred)

In [ ]:
cm_k

array([[ 0,  0,  1,  1,  0,  0],
       [ 0,  1,  4,  5,  0,  0],
       [ 0,  4, 96, 39,  2,  0],
       [ 0,  1, 51, 74, 10,  1],
       [ 0,  1,  6,  9, 11,  0],
       [ 0,  0,  0,  1,  2,  0]])

In [ ]:
ac_k

0.56875

##Support Vector Machines

In [ ]:
from sklearn.svm import SVC
scla = SVC(kernel='linear')
scla.fit(X_train,y_train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='linear',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [ ]:
y_pred = scla.predict(X_test)

In [ ]:
from sklearn.metrics import accuracy_score, confusion_matrix
cm_s = confusion_matrix(y_test, y_pred)
ac_s = accuracy_score(y_test,y_pred)
print(cm_s)
print(ac_s)

[[  0   0   2   0   0   0]
 [  0   0   8   2   0   0]
 [  0   0 108  33   0   0]
 [  0   0  57  66  14   0]
 [  0   0   3  18   6   0]
 [  0   0   0   1   2   0]]
0.5625


## Accuracy score

In [ ]:
print(ac_l, ac_d , ac_r , ac_k , ac_s)

0.578125 0.590625 0.68125 0.56875 0.5625


As Random Forest Classification gives the highest accuracy score of 68.1% it is the best fit model.